oauth2/src/test/java/com/linecorp/armeria/client/auth/oauth2/OAuth2ClientCredentialsGrantTest.java

```diff
@@ -0,0 +1,209 @@
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.auth.oauth2;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ForkJoinPool;
+import java.util.stream.Collectors;
+
+import org.junit.Rule;
+import org.junit.Test;
+
+import com.google.common.collect.ImmutableMap;
+
+import com.linecorp.armeria.client.WebClient;
+import com.linecorp.armeria.common.AggregatedHttpResponse;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.auth.oauth2.InvalidClientException;
+import com.linecorp.armeria.common.auth.oauth2.MockOAuth2AccessToken;
+import com.linecorp.armeria.common.auth.oauth2.MockOAuth2ClientCredentialsService;
+import com.linecorp.armeria.common.auth.oauth2.MockOAuth2IntrospectionService;
+import com.linecorp.armeria.server.AbstractHttpService;
+import com.linecorp.armeria.server.HttpService;
+import com.linecorp.armeria.server.Server;
+import com.linecorp.armeria.server.ServerBuilder;
+import com.linecorp.armeria.server.ServiceRequestContext;
+import com.linecorp.armeria.server.auth.oauth2.OAuth2TokenIntrospectionAuthorizer;
+import com.linecorp.armeria.testing.junit4.server.ServerRule;
+
+public class OAuth2ClientCredentialsGrantTest {
+
+    static final String CLIENT_CREDENTIALS = "dGVzdF9jbGllbnQ6Y2xpZW50X3NlY3JldA=="; //test_client:client_secret
+    static final String SERVER_CREDENTIALS = "dGVzdF9zZXJ2ZXI6c2VydmVyX3NlY3JldA=="; //test_server:server_secret
+
+    static final MockOAuth2AccessToken TOKEN =
+        MockOAuth2AccessToken.generate("test_client", null, Duration.ofHours(3L),
+                                       ImmutableMap.of("extension_field", "twenty-seven"), "read", "write");
+
+    static final HttpService SERVICE = new AbstractHttpService() {
+        @Override
+        protected HttpResponse doGet(ServiceRequestContext ctx, HttpRequest req)
+                throws Exception {
+            return HttpResponse.of(HttpStatus.OK);
+        }
+    };
+
+    @Rule
+    public ServerRule authServerRule = new ServerRule() {
```

Comment:
Could you use JUnit 5 and `ServerExtension`? We prefer JUnit 5 for the new code.
https://armeria.dev/community/developer-guide#use-junit-5-instead-of-junit-4-for-testing

Reply:
Migrated to `ServerExtension` and JUnit 5.